File: MobMuPlatEditor-OSXLionAndUp/MobMuPlatEditor/Controls/MMPGrid.m
//
//  MMPGrid.m
//  MobMuPlatEd1
//
//  Created by Daniel Iglesia on 1/1/13.
//  Copyright (c) 2013 Daniel Iglesia. All rights reserved.
//

#import "MMPGrid.h"

@implementation MMPGrid {
  NSTrackingArea *currentTrackingArea;
  NSControl* heldControl;//used in mode 1+2 for detecting when we have dragged out of a control.
}

- (id)initWithFrame:(NSRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        self.address=@"/myGrid";
        gridButtons = [[NSMutableArray alloc]init];
        //defaults
        [self setCellPadding:2];
        [self setBorderThickness:3];
        [self setDimX:4];
        [self setDimY:3];
        [self redrawDim];
        //[self setFrame:frame];//necc?
        
        [self addHandles];
    }
    return self;
}

-(void)setModeObjectUndoable:(NSNumber*)inVal{
  [[self undoManager] registerUndoWithTarget:self selector:@selector(setModeObjectUndoable:) object:[NSNumber numberWithInt:[self mode]]];
  
  [self setMode:[inVal intValue]];
}


-(void)setCellPaddingUndoable:(NSNumber*)inVal{
    [[self undoManager] registerUndoWithTarget:self selector:@selector(setCellPaddingUndoable:) object:[NSNumber numberWithInt:[self cellPadding]]];
    
    [self setCellPadding:[inVal intValue]];
}

-(void)setCellPadding:(int)cellPadding{
    _cellPadding=cellPadding;
    [self redrawDim];
}

-(void)setBorderThicknessUndoable:(NSNumber*)inVal{
    [[self undoManager] registerUndoWithTarget:self selector:@selector(setBorderThicknessUndoable:) object:[NSNumber numberWithInt:[self borderThickness]]];
    
    [self setBorderThickness:[inVal intValue]];
}

-(void)setBorderThickness:(int)borderThickness{
    _borderThickness = borderThickness;
    for(NSControl* button in gridButtons)button.layer.borderWidth=_borderThickness;
}

-(void)hackRefresh{
    [super hackRefresh];
    for(NSControl* button in gridButtons){
        button.layer.borderWidth=_borderThickness;
      if(button.tag==1)button.layer.backgroundColor=self.highlightColor.CGColor;
      else button.layer.backgroundColor=[[NSColor clearColor] CGColor];
    }
}


-(void)setFrame:(NSRect)frameRect{
    [super setFrame:frameRect];
    
    float buttonWidth = self.frame.size.width/_dimX;
    float buttonHeight = self.frame.size.height/_dimY;
    
    for(int j=0;j<_dimY;j++){
        for(int i=0;i<_dimX;i++){
            NSView* buttonView = [gridButtons objectAtIndex:j*_dimX+i ];
            [buttonView setFrame:CGRectMake(i*buttonWidth, j*buttonHeight, buttonWidth-_cellPadding, buttonHeight-_cellPadding)];
        }
    }
}

-(void)setDimXUndoable:(NSNumber*)inVal{
    [[self undoManager] registerUndoWithTarget:self selector:@selector(setDimXUndoable:) object:[NSNumber numberWithInt:[self dimX]]];
    
    [self setDimX:[inVal intValue]];
}

-(void)setDimX:(int)inX{
    _dimX=inX;
    [self redrawDim];
}

-(void)setDimYUndoable:(NSNumber*)inVal{
    [[self undoManager] registerUndoWithTarget:self selector:@selector(setDimYUndoable:) object:[NSNumber numberWithInt:[self dimY]]];
    
    [self setDimY:[inVal intValue]];
}

-(void)setDimY:(int)inY{
    _dimY=inY;
    [self redrawDim];
}

//whenever dim x or y has changed, redraw everything
-(void)redrawDim{
    
    float buttonWidth = self.frame.size.width/_dimX;
    float buttonHeight = self.frame.size.height/_dimY;
    
    for(NSControl* button in gridButtons){
        [button removeFromSuperview];
    }
    [gridButtons removeAllObjects];
    
    for(int j=0;j<_dimY;j++){
        for(int i=0;i<_dimX;i++){
          CGRect rect = CGRectMake(i*buttonWidth, j*buttonHeight, buttonWidth-_cellPadding, buttonHeight-_cellPadding);
            NSControl* newButtonView = [[NSControl alloc]initWithFrame:rect];
            [newButtonView setWantsLayer:YES];
            newButtonView.layer.cornerRadius=2;
            newButtonView.layer.borderWidth=_borderThickness;
          newButtonView.layer.borderColor=self.color.CGColor;
            [gridButtons addObject:newButtonView];
            [self addSubview:newButtonView];
          
          NSTrackingArea* trackingArea = [[NSTrackingArea alloc] initWithRect:newButtonView.bounds options:NSTrackingMouseEnteredAndExited | NSTrackingActiveAlways|NSTrackingInVisibleRect owner:newButtonView userInfo:nil];
              [newButtonView addTrackingArea:trackingArea];
            
        }
    }
    //bring handle up front
    [self bringUpHandle];
    
}

-(void)mouseDown:(NSEvent *)event
{//NSLog(@"currtrack %@", currentTrackingArea.owner);
    [super mouseDown:event];
    if(![self.editingDelegate isEditing] && currentTrackingArea){
      NSControl *currControl = currentTrackingArea.owner;
      
      if(_mode==0){//toggle on down
        if(currControl.tag==0)[self doOn:currControl];
        else [self doOff:currControl];
      }
      else if (_mode==1){//momentary
        heldControl = currControl;
        if(currControl.tag==0)[self doOn:currControl];
      }
      else {//hybrid, on change
        heldControl = currControl;
        if(currControl.tag==0)[self doOn:currControl];
        else [self doOff:currControl];
      }
    }
}

-(void)mouseExited:(NSEvent *)theEvent{
  [super mouseExited:theEvent];
  if(_mode==1) {//release button if it was the one that was just pressed
    if([[theEvent trackingArea] owner] == heldControl) {
      [self doOff:heldControl];
      heldControl=nil;
    }
  }
  
  currentTrackingArea = nil;
}

-(void)mouseEntered:(NSEvent *)theEvent{
  [super mouseEntered:theEvent];
  currentTrackingArea = theEvent.trackingArea;
}

-(void)mouseUp:(NSEvent *)event{

  [super mouseUp:event];
  if(![self.editingDelegate isEditing] && currentTrackingArea){
    
    NSControl *currControl = currentTrackingArea.owner;
    
    if(_mode ==1 && currControl.tag==1 &&currControl == heldControl)
      [self doOff:currControl];
    else if (_mode ==2 && currControl.tag==1 && currControl==heldControl)
      [self doOff:currControl];
        
    heldControl=nil;
  }
}

-(void)doOn:(NSControl*)control{
  [control setTag:1];
  control.layer.backgroundColor=self.highlightColor.CGColor;
  
  [self sendValueForControl:control];
}

- (void)doOff:(NSControl*)control{
  [control setTag:0];
  control.layer.backgroundColor=[[NSColor clearColor]CGColor];
  
 [self sendValueForControl:control];
}

//- (void)sendValueX:(int)x Y:(int)y val:(int)val {
-(void)sendValueForControl:(NSControl*)control{
  
  //get coordinate of cell
  int controlIndex = (int)[gridButtons indexOfObject:control];
  
//send message
  NSMutableArray* formattedMessageArray = [[NSMutableArray alloc]init];
  [formattedMessageArray addObject:self.address];
  [formattedMessageArray addObject:[NSNumber numberWithInt:controlIndex%_dimX]];
  [formattedMessageArray addObject:[NSNumber numberWithInt:controlIndex/_dimX]];
  [formattedMessageArray addObject:[NSNumber numberWithInt:(int)control.tag]];
  [self.editingDelegate sendFormattedMessageArray:formattedMessageArray];
}

-(void)setColor:(NSColor *)color{
    [super setColor:color];
    for(NSControl* hitView in gridButtons)hitView.layer.borderColor=color.CGColor;
}

-(void)setHighlightColor:(NSColor *)color{
    [super setHighlightColor:color];
    for(NSControl* hitView in gridButtons)if(hitView.tag==1)hitView.layer.backgroundColor=color.CGColor;
}

//receive messages from PureData (via [send toGUI], routed through the PdWrapper.pd patch), routed from Document via the address to this object

-(void)receiveList:(NSArray *)inArray{
    
    BOOL sendVal=YES;
    //if message preceded by "set", then set "sendVal" flag to NO, and strip off set and make new messages array without it
    if ([inArray count]>0 && [[inArray objectAtIndex:0] isKindOfClass:[NSString class]] && [[inArray objectAtIndex:0] isEqualToString:@"set"]){
        NSRange newRange = (NSRange){1, [inArray count]-1};
        inArray = [inArray subarrayWithRange: newRange];
        //printf("\nset!");
        sendVal=NO;
    }
    
    //if message is three numbers, look at message and set my value, outputting value if required
    if([inArray count]==3 && [[inArray objectAtIndex:0] isKindOfClass:[NSNumber class]] && [[inArray objectAtIndex:1] isKindOfClass:[NSNumber class]] && [[inArray objectAtIndex:2] isKindOfClass:[NSNumber class]]){
        int indexX = (int)[[inArray objectAtIndex:0] floatValue];
        int indexY = (int)[[inArray objectAtIndex:1] floatValue];
        int val = (int)[[inArray objectAtIndex:2] floatValue];
        if(indexX<_dimX && indexY<_dimY){
            
            NSControl* currButton = [gridButtons objectAtIndex:indexX+indexY*_dimX];
            if(val>1)val=1;if(val<0)val=0;
            currButton.tag=val;
            if(val==0)currButton.layer.backgroundColor=[[NSColor clearColor] CGColor];
            else currButton.layer.backgroundColor=self.highlightColor.CGColor;
            
            if(sendVal){
                NSMutableArray *formattedMessageArray = [NSMutableArray arrayWithObjects:self.address, [NSNumber numberWithInt:indexX], [NSNumber numberWithInt:indexY], [NSNumber numberWithInt:val],  nil];
                [self.editingDelegate sendFormattedMessageArray:formattedMessageArray];
            }
        }
    }
    
    //else if message starts with "getColumn", spit out array of that column's values
    else if([inArray count]==2 && [[inArray objectAtIndex:0] isKindOfClass:[NSString class]] && [[inArray objectAtIndex:0] isEqualToString:@"getcolumn"] && [[inArray objectAtIndex:1] isKindOfClass:[NSNumber class]]){
        int colIndex = [[inArray objectAtIndex:1] intValue];
        //printf("\n getcol %d", colIndex);
        if(colIndex>=0 && colIndex<_dimX){
            NSMutableArray *formattedMessageArray = [NSMutableArray arrayWithObjects:self.address,@"column", nil];
        
            for(int i=0;i<_dimY;i++){
                int val = (int)[[gridButtons objectAtIndex:(colIndex+_dimX*i)] tag];//0 or 1
                [formattedMessageArray addObject:[NSNumber numberWithInt:val]];
            }
          [self.editingDelegate sendFormattedMessageArray:formattedMessageArray];
        }
    }
    
    //else if message starts with "getRow", spit out array of that row's values
    else if([inArray count]==2 && [[inArray objectAtIndex:0] isKindOfClass:[NSString class]] && [[inArray objectAtIndex:0] isEqualToString:@"getrow"] && [[inArray objectAtIndex:1] isKindOfClass:[NSNumber class]]){
        int rowIndex = [[inArray objectAtIndex:1] intValue];
        //printf("\n getrow %d dimX %d: ", rowIndex, _dimX);
        if(rowIndex>=0 && rowIndex<_dimY){
            NSMutableArray *formattedMessageArray = [NSMutableArray arrayWithObjects:self.address, @"row", nil];
        
            for(int i=0;i<_dimX;i++){
                int val = (int)[[gridButtons objectAtIndex:(i+_dimX*rowIndex)] tag];//0 or 1
                [formattedMessageArray addObject:[NSNumber numberWithInt:val]];
            }
            [self.editingDelegate sendFormattedMessageArray:formattedMessageArray];
        }
    }
    //clear
    else if([inArray count]==1 && [[inArray objectAtIndex:0] isKindOfClass:[NSString class]] && [[inArray objectAtIndex:0] isEqualToString:@"clear"]){
        for(NSControl* currButton in gridButtons){
            currButton.tag=0;
            currButton.layer.backgroundColor=[[NSColor clearColor]CGColor];
        }
    }
}

//coder for copy/paste

- (void)encodeWithCoder:(NSCoder *)coder {
    [super encodeWithCoder:coder];
    [coder encodeInt:self.dimX forKey:@"dimX"];
    [coder encodeInt:self.dimY forKey:@"dimY"];
    [coder encodeInt:self.borderThickness forKey:@"borderThickness"];
    [coder encodeInt:self.cellPadding forKey:@"cellPadding"];
    
}

- (id)initWithCoder:(NSCoder *)coder {
    
    if(self=[super initWithCoder:coder]){
        [self setDimX:[coder decodeIntForKey:@"dimX"]];
        [self setDimY:[coder decodeIntForKey:@"dimY"]];
        [self setBorderThickness:[coder decodeIntForKey:@"borderThickness"]];
        [self setCellPadding:[coder decodeIntForKey:@"cellPadding"]];
    }
    return self;
}




@end

File: MobMuPlatEditor-Java/src/com/iglesiaintermedia/MobMuPlatEditor/controls/MMPGrid.java
package com.iglesiaintermedia.MobMuPlatEditor.controls;

import java.awt.BasicStroke;
import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.Rectangle;
import java.awt.RenderingHints;
import java.awt.event.MouseEvent;
import java.util.ArrayList;

import javax.swing.JPanel;

public class MMPGrid extends MMPControl {
	static final int EDGE_RADIUS = 2;
	static final Color clearColor = new Color(0,0,0,0);
	
	public int dimX;
	public int dimY;
	public int borderThickness;
	public int cellPadding;
	int mode;
	
	
	public void setMode(int mode){
		this.mode = mode;
	}
	public int getMode(){
		return mode;
	}
	
	ArrayList<TogglePanel> togglePanelArray; 
	
	public MMPGrid(MMPGrid otherGrid){
		this(otherGrid.getBounds());//normal constructor
		this.setColor(otherGrid.color);
		this.setHighlightColor(otherGrid.highlightColor);
		this.address=otherGrid.address;
		
		this.setMode(otherGrid.mode);
		this.setDimX(otherGrid.dimX);
		this.setDimY(otherGrid.dimY);
		this.setBorderThickness(otherGrid.borderThickness);
		this.setCellPadding(otherGrid.cellPadding);
		
		
	}
	
	public MMPGrid(Rectangle frame){
		super();
		//borderThickness=5;
		
		togglePanelArray = new ArrayList<TogglePanel>();
		dimY=2;
		dimX=2;
		address="/myGrid";
		
		
		setCellPadding(2);
		setBorderThickness(3);
		setDimX(4);
		setDimY(3);
		
		
		this.addMouseListener(this);
		this.addMouseMotionListener(this);
		this.setColor(this.color);
		this.setBounds(frame);
		
	}
	
	public void setBounds(Rectangle frame){
		super.setBounds(frame);
		
		int buttonWidth = getWidth()/dimX;
		int buttonHeight = getHeight()/dimY;
		
		for(int j=0;j<dimY;j++){
	        for(int i=0;i<dimX;i++){
			TogglePanel currTogglePanel = togglePanelArray.get(j*dimX+i);
			currTogglePanel.setBounds(new Rectangle(i*buttonWidth, j*buttonHeight, buttonWidth-cellPadding, buttonHeight-cellPadding));
	        }
		}
	}
	
	public void setCellPadding(int inPadding){
		cellPadding = inPadding;
		redrawDim();
	}
	public void setBorderThickness(int inThickness){
		borderThickness = inThickness;
		redrawDim();//ness whole redraw??? just iterate and set thickness
	}
	public void setDimX(int inDimX){
		dimX = inDimX;
		redrawDim();
	}
	public void setDimY(int inDimY){
		dimY = inDimY;
		redrawDim();
	}
	
	void redrawDim(){
		int buttonWidth = getWidth()/dimX;
		int buttonHeight = getHeight()/dimY;
		//System.out.print("\nredraw dim "+ buttonWidth+" "+buttonHeight);
		
		for(TogglePanel tp : togglePanelArray)tp.getParent().remove(tp);
		togglePanelArray.clear();
		
		 for(int j=0;j<dimY;j++){
		        for(int i=0;i<dimX;i++){
		            TogglePanel togglePanel = new TogglePanel();
		            togglePanel.setBorderThickness(borderThickness);
		            togglePanel.addMouseListener(this);
		            togglePanel.addMouseMotionListener(this);
		            togglePanelArray.add(togglePanel);
		            add(togglePanel);
		            //for some reason this only causes refresh if done after adding to panel
		            togglePanel.setBounds(new Rectangle(i*buttonWidth, j*buttonHeight, buttonWidth-cellPadding, buttonHeight-cellPadding));
	        }
		   }
		 
		 this.repaint();//helps...

	}
	
	public void setColor(Color newColor){
		super.setColor(newColor);
		
		
		this.repaint();//repaint border
	}
	
	/*public void mousePressed(MouseEvent e) {
		//System.out.print(e.getClickCount() + " click(s)");
		super.mousePressed(e);
		
		if(!editingDelegate.isEditing()){
	      //setValue(1-value);
	      //sendValue();
			for(TogglePanel tp: togglePanelArray){
				if(tp.getBounds().contains(e.getX(), e.getY())){
					//System.out.print("\nhit togglepanel "+togglePanelArray.indexOf(tp));
					tp.value = 1-tp.value;
					if(tp.value==1)tp.setColor(highlightColor);
					else tp.setColor(clearColor);
					
					int hitViewIndex = togglePanelArray.indexOf(tp);
					
					Object[] args = new Object[]{new Integer(hitViewIndex%dimX), new Integer(hitViewIndex/dimX), new Integer(tp.value)};
					editingDelegate.sendMessage(this.address, args);
				}
			}
	      
	      
	    }
	}*/
	
	void sendValueOfPanel(TogglePanel tp){
		int hitViewIndex = togglePanelArray.indexOf(tp);
		
		Object[] args = new Object[]{new Integer(hitViewIndex%dimX), new Integer(hitViewIndex/dimX), new Integer(tp.value)};
		editingDelegate.sendMessage(this.address, args);
	}
	
	void doOn(TogglePanel tp){
		tp.value=1;
		tp.setColor(highlightColor);
		sendValueOfPanel(tp);
	}
	void doOff(TogglePanel tp){
		tp.value=0;
		tp.setColor(clearColor);
		sendValueOfPanel(tp);
	}
	
	public void mouseDragged(MouseEvent e) {
		super.mouseDragged(e);
		if(!editingDelegate.isEditing() && e.getComponent()!=this){
	    	if(mode==1) {//release button if it was the one that was just pressed
	    		TogglePanel tp = (TogglePanel) e.getComponent();
	    	    if(!tp.contains(e.getPoint()) && tp.value==1) {
	    	      doOff(tp);
	    	    }
	    	  }
	    	}
	}
	
	public void mousePressed(MouseEvent e){
		super.mousePressed(e);
		
		if(!editingDelegate.isEditing() && e.getComponent()!=this){
			//System.out.print("\nhit togglepanel");
			TogglePanel tp = (TogglePanel) e.getComponent();
			if(mode==0){
				if(tp.value==0)doOn(tp);
				else doOff(tp);
			} 
			else if (mode==1){//momentary
		        if(tp.value==0)doOn(tp);
		    }
		    else {//hybrid, on change
		        if(tp.value==0)doOn(tp);
		        else doOff(tp);
		    }
		}
	}
	public void mouseReleased(MouseEvent e) {
		super.mouseReleased(e);
			
		if(!editingDelegate.isEditing() && e.getComponent()!=this){
			TogglePanel tp = (TogglePanel) e.getComponent();
				
			if(mode ==1 && tp.value==1){
				doOff(tp);
			}
			else if (mode ==2 && tp.value==1 && tp.contains(e.getPoint())){//if still within the panel, off
				doOff(tp);
			}
		}
    }

	class TogglePanel extends JPanel{
		RoundedBorderPanel borderPanel;
		RoundedPanel touchPanel;
		int value;
		public TogglePanel(){
			super();
			setLayout(null);
			borderPanel = new RoundedBorderPanel();
			add(borderPanel);
			
			touchPanel = new RoundedPanel();
			touchPanel.setCornerRadius(EDGE_RADIUS);
			touchPanel.setBackground(clearColor);
			add(touchPanel);
		
			setOpaque(false);
		}
		
		public void setColor(Color inColor){
			touchPanel.setBackground(inColor);
		}
		
		
		public void setBounds(Rectangle frame){
			super.setBounds(frame);
			borderPanel.setBounds(new Rectangle(0,0,this.getWidth(), this.getHeight()));
			//touchPanel.setBounds(new Rectangle(0,0,this.getWidth(), this.getHeight()));
			touchPanel.setBounds(borderThickness/2, borderThickness/2, getWidth()-borderThickness, getHeight()-borderThickness);

		}
		
		public void setBorderThickness(int inBorderThickness){
			//borderPanel.setBorderThickness(inBorderThickness);
			this.setBounds(this.getBounds());
		}
	
		
	}
	
	class RoundedBorderPanel extends JPanel{
		//int borderThickness;
		//protected Dimension arcs = new Dimension(EDGE_RADIUS, EDGE_RADIUS);
		
		public RoundedBorderPanel(){
			super();
			setOpaque(false);
			//borderThickness=5;
		}
		
		/*public void setBorderThickness(int inBorderThickness){
			borderThickness = inBorderThickness;
		}*/
		
		
		protected void paintBorder(Graphics g) {
	        //System.out.print("\npaintBorder w h "+getWidth()+" "+this.getHeight());
			if(borderThickness>0){//added this because on borderthickness=0 it was still drawing a border...
	        Graphics2D g2 = (Graphics2D)g.create();
	        g2.setColor(color);
	        g2.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);
	        g2.setStroke(new BasicStroke(borderThickness,BasicStroke.CAP_ROUND, BasicStroke.JOIN_ROUND));
	        g2.drawRoundRect(borderThickness/2, borderThickness/2, getWidth()-borderThickness-1, this.getHeight()-borderThickness-1, EDGE_RADIUS*2, EDGE_RADIUS*2);
			}
		}
		
		
		
	}
	
	//receive messages from PureData (via [send toGUI], routed through the PdWrapper.pd patch), routed from Document via the address to this object
	public void receiveList(ArrayList<Object> messageArray){
		//System.out.print("receive size "+messageArray.size()+" element 1 int? "+(messageArray.get(1) instanceof Integer));
		boolean sendVal  = true;
		//if message preceded by "set", then set "sendVal" flag to NO, and strip off set and make new messages array without it
	    if (messageArray.size()>0 && (messageArray.get(0) instanceof String) && messageArray.get(0).equals("set") ){
	        //messageArray.remove(0);bad!
	    	messageArray = new ArrayList<Object>(messageArray.subList(1, messageArray.size() ) );
	        sendVal=false;
	    }
	    
	    //if message is three numbers, look at message and set my value, outputting value if required
	    if (messageArray.size()==3 && (messageArray.get(0) instanceof Float || messageArray.get(0) instanceof Integer) && (messageArray.get(1) instanceof Float || messageArray.get(1) instanceof Integer) && (messageArray.get(2) instanceof Float || messageArray.get(2) instanceof Integer)){
	 	   int indexX =0;
	 	   if(messageArray.get(0) instanceof Float)  indexX= (int) ((Float)(messageArray.get(0))).floatValue() ;
	 	   else indexX=  ((Integer)(messageArray.get(0))).intValue() ;
	 	   
	 	  int indexY =0;
	 	   if(messageArray.get(1) instanceof Float)  indexY= (int) ((Float)(messageArray.get(0))).floatValue() ;
	 	   else indexY=  ((Integer)(messageArray.get(1))).intValue() ;
	 	   
	 	  int val =0;
	 	   if(messageArray.get(2) instanceof Float)  val= (int) ((Float)(messageArray.get(0))).floatValue() ;
	 	   else val=  ((Integer)(messageArray.get(2))).intValue() ;
	 	   
	 	 if(indexX<dimX && indexY<dimY){
	 		 TogglePanel currTogglePanel = togglePanelArray.get(indexX+indexY*dimX);
	 		if(val>0)val=1;if(val<0)val=0;
	 		currTogglePanel.value=val;
	 		if(val==1)currTogglePanel.setColor(highlightColor);
			else currTogglePanel.setColor(clearColor);
	 		
	 		if(sendVal){
	 			Object[] args = new Object[]{new Integer(indexX), new Integer(indexY), new Integer(val)};
				editingDelegate.sendMessage(this.address, args);
	 		}
	 	 }
	    }
	 	 
	 	//else if message starts with "getColumn", spit out array of that column's values
	 	 else if (messageArray.size()==2 && (messageArray.get(0) instanceof String) && (messageArray.get(0).equals("getcolumn")) && (messageArray.get(1) instanceof Float || messageArray.get(1) instanceof Integer) ) {
		 	   int colIndex =0;
		 	  if(messageArray.get(1) instanceof Float)  colIndex= (int) ((Float)(messageArray.get(1))).floatValue() ;
		 	   else colIndex=  ((Integer)(messageArray.get(1))).intValue() ;
		 	  
		 	   if(colIndex>=0 && colIndex<dimX){
		 		   Object[] args = new Object[dimY];
		 		  for(int i=0;i<dimY;i++){
		 			  int currValue = togglePanelArray.get(colIndex+dimX*i).value;
		 			  args[i]=new Integer(currValue);
		 		  }
		 		 editingDelegate.sendMessage(this.address, args);
		 	   }
	 	 }
	 	//else if message starts with "getRow", spit out array of that row's values
	 	else if (messageArray.size()==2 && (messageArray.get(0) instanceof String) && (messageArray.get(0).equals("getrow")) && (messageArray.get(1) instanceof Float || messageArray.get(1) instanceof Integer) ) {
	 		 int rowIndex =0;
		 	  if(messageArray.get(1) instanceof Float)  rowIndex= (int) ((Float)(messageArray.get(1))).floatValue() ;
		 	   else rowIndex=  ((Integer)(messageArray.get(1))).intValue() ;
		 	  
		 	  if(rowIndex>=0 && rowIndex<dimY){
		 		   Object[] args = new Object[dimY];
		 		  for(int i=0;i<dimX;i++){
		 			  int currValue = togglePanelArray.get(i+dimX*rowIndex).value;
		 			  args[i]=new Integer(currValue);
		 		  }
		 		 editingDelegate.sendMessage(this.address, args);
		 	   }
	 	 }
	 	//clear
	 	else if (messageArray.size()==1 && (messageArray.get(0) instanceof String) && (messageArray.get(0).equals("clear")) ) {
			for(TogglePanel tp : togglePanelArray){
				if(tp.value==1){
					tp.value=0;
					tp.setColor(clearColor);
				}
			}
			//this.repaint();
	 	}
	    
	    
	    

	}
}
